### ISSUE TITLE
BUG: unexpected broadening following `isinstance(x, Sequence)` and `isinstance(x, Mapping)`

### ISSUE THREAD

**Issue by neutrinoceros** [astral-sh/ty#3684]
### Summary

The following playground illustrates a minified self-contained version of a string processing library I maintain, where I see what I think are 3 false positives that probably are symptoms of a common defect or limitation. In essence, I see unexpected broadening of inferred types following branching on `isinstance(x, Sequence)` or `isinstance(x, Mapping)`. In all cases, the type(s) of contained objects are known, but seemingly lost (degraded to `object`) post branching
https://play.ty.dev/4a5c7907-c6a0-4fa2-a798-4fe70c648fc0

I see the same problem with this minimal snippet
```py
from collections.abc import Sequence

def first(obj: int | Sequence[int], /) -> int:
    if isinstance(obj, Sequence):
        return obj[0] # false positive
    else:
        return obj
```

### Version

ty 0.0.43 (29ce3146f 2026-06-03)

**Comment by neutrinoceros** [comment]
it seems unlikely to me that this wasn't reported already but I wasn't able to find an existing report for it. Apologies if it's indeed a dupe !

**Comment by neutrinoceros** [comment]
FWIW pyrefly reports no errors for these examples, even with `--preset=strict`

**Comment by neutrinoceros** [comment]
Possibly a duplicate of, or subissue to #3557 ?

**Comment by rzwq** [comment]
you might want to use `issubclass(obj, Sequence)`?

**Comment by neutrinoceros** [comment]
No I'm really checking the type of instances, not class hierarchies.

**Comment by sharkdp** [comment]
In your minimal snippet, ty accounts for the possibility of a common subclass of `Sequence` and `int`. So when you check for `Sequence` instances, you can't be sure that the sequences actually contain `int`s. See https://docs.astral.sh/ty/reference/typing-faq/#what-is-toplistunknown-and-why-does-it-appear for more context.

You can fix this by doing `if not isinstance(.., int)` instead of `if isinstance(.., Sequence)`.

**Comment by neutrinoceros** [comment]
I see. Just to check, this means that `int` and `Sequence` are not _disjoint_, right ?
Inverting my conditions works for the first real-life example but not for the second. At least not trivially. I'll try a bit harder later and report back here (possibly just closing the issue).

**Comment by AlexWaygood** [comment]
> I see. Just to check, this means that `int` and `Sequence` are not _disjoint_, right ?

That's correct. It's perfectly possible to create a class at runtime that inherits from both. (This is _not_ the case for certain builtin classes such as `tuple` and `list`, and ty is aware of this.)

```pycon
>>> from collections.abc import Sequence
>>> class Foo(int, Sequence[str]): ...
...
>>>
```

**Comment by neutrinoceros** [comment]
Sorry guys, even with the previous explanation I still cannot make sense of the remaining diagnostics...
Here's an updated version of my playground MRE, where the first diagnostic has been addressed. It's probably still not minimal but I don't really see how to simplify it :/
https://play.ty.dev/1222577c-0f77-4910-b2b0-5bea5c478d29

**Comment by neutrinoceros** [comment]
(I'll try to teach myself more type theory so I don't constantly run to your bug tracker for guidance)

**Comment by carljm** [comment]
@neutrinoceros It's totally fine to bring issues to the bug tracker if they don't make sense to you! Even if ty is correct, it gives us good signal on what behaviors are confusing and might need adjusting, or better diagnostics.

And this remaining issue is a good case in point, because it really is a ty bug: this example suffers from https://github.com/astral-sh/ty/issues/3557 -- see the last example there specifically mentions `enumerate`. Sorry for the buggy behavior, I have a branch in progress locally to fix this!
